options/ansi/generic/string-stubs.cpp

```diff
@@ -414,7 +414,144 @@ int strverscmp(const char *l0, const char *r0) {
 	return l[i] - r[i];
 }
 
-void *memmem(const void *, size_t, const void *, size_t) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+// Taken from musl.
+static char *twobyte_memmem(const unsigned char *h, size_t k, const unsigned char *n) {
+	uint16_t nw = n[0]<<8 | n[1], hw = h[0]<<8 | h[1];
+	for (h+=2, k-=2; k; k--, hw = hw<<8 | *h++)
+		if (hw == nw) return (char *)h-2;
+	return hw == nw ? (char *)h-2 : 0;
+}
+
+static char *threebyte_memmem(const unsigned char *h, size_t k, const unsigned char *n) {
+	uint32_t nw = (uint32_t)n[0]<<24 | n[1]<<16 | n[2]<<8;
+	uint32_t hw = (uint32_t)h[0]<<24 | h[1]<<16 | h[2]<<8;
+	for (h+=3, k-=3; k; k--, hw = (hw|*h++)<<8)
+		if (hw == nw) return (char *)h-3;
+	return hw == nw ? (char *)h-3 : 0;
+}
+
+static char *fourbyte_memmem(const unsigned char *h, size_t k, const unsigned char *n) {
+	uint32_t nw = (uint32_t)n[0]<<24 | n[1]<<16 | n[2]<<8 | n[3];
+	uint32_t hw = (uint32_t)h[0]<<24 | h[1]<<16 | h[2]<<8 | h[3];
+	for (h+=4, k-=4; k; k--, hw = hw<<8 | *h++)
+		if (hw == nw) return (char *)h-4;
+	return hw == nw ? (char *)h-4 : 0;
+}
+
+#define MAX(a,b) ((a)>(b)?(a):(b))
+#define MIN(a,b) ((a)<(b)?(a):(b))
+
+#define BITOP(a,b,op) \
```

Comment:
This macro is also quite unreadable. Generally we don't like to use macros. This could technically also be replaced by a lambda templated on the operation, but that might need some basic additions to cxxshim (like ``std::bit_or`` and ``std::bit_and``, they aren't very hard to implement though).